packages/vite/src/node/config.ts

```diff
@@ -792,7 +792,11 @@ function resolveDepOptimizationOptions(
     exclude: optimizeDeps.exclude ?? [],
     needsInterop: optimizeDeps.needsInterop ?? [],
     extensions: optimizeDeps.extensions ?? [],
-    noDiscovery: optimizeDeps.noDiscovery ?? consumer !== 'client',
+    noDiscovery:
+      optimizeDeps.noDiscovery ??
+      // If the consumer is not the client env (which default discovery entrypoint is `index.html`)
+      // and there are no entries from where to start crawling then disable discovery
+      (consumer !== 'client' && !optimizeDeps.entries?.length),
```

Comment:
I think it would be better to let server environments explicitly opt-in here setting `noDiscovery: false` to avoid making the default more complex.

Reply:
Thanks for the suggestion, it makes sense to me 🙂

I've updated this 👍